### ISSUE TITLE
[CUDA] Tweedie metric receives raw log-space scores without exp() conversion, breaking early stopping

### ISSUE THREAD

**Issue by Whisdangist** [microsoft/LightGBM#7152]
## Description

When using `device='cuda'` with `objective='tweedie'` and `metric='tweedie'`, the reported tweedie metric values are **incorrect**, causing early stopping to malfunction. The CUDA tweedie metric receives raw log-space scores without the required `exp()` transformation, producing nonsensical loss values.

This happens because the tweedie **objective** has no CUDA implementation and falls back to the CPU class `RegressionTweedieLoss`, while the tweedie **metric** does have a CUDA implementation (`CUDATweedieMetric`). The CPU fallback objective does not override `NeedConvertOutputCUDA()`, so the CUDA metric skips the `exp()` score conversion and evaluates on raw log-space scores instead of natural-space predictions.

## Reproducing

```python
import numpy as np
import lightgbm as lgb

np.random.seed(42)
n = 5000
X = np.random.randn(n, 10).astype(np.float32)
y = np.clip(np.abs(np.exp(0.3 * X[:, 0]) + np.random.exponential(1.0, n)), 0.01, 100.0)

split = int(n * 0.8)
X_tr, X_te = X[:split], X[split:]
y_tr, y_te = y[:split], y[split:]

params = {
    'objective': 'tweedie', 'metric': 'tweedie',
    'tweedie_variance_power': 1.5,
    'learning_rate': 0.1, 'num_leaves': 31,
    'min_data_in_leaf': 20, 'verbose': -1, 'seed': 42,
}

for device in ['cpu', 'cuda']:
    p = {**params, 'device': device}
    train_ds = lgb.Dataset(X_tr, label=y_tr, free_raw_data=False)
    val_ds = lgb.Dataset(X_te, label=y_te, reference=train_ds, free_raw_data=False)
    evals = {}
    model = lgb.train(p, train_ds, num_boost_round=5,
                      valid_sets=[val_ds], valid_names=['valid'],
                      callbacks=[lgb.record_evaluation(evals)])
    print(f"[{device.upper()}] {evals['valid']['tweedie']}")
```

**Expected output** (both should be similar ~5.7):
```
[CPU]  [5.7251, 5.7227, 5.7208, 5.7193, 5.7179]
[CUDA] [5.72xx, 5.72xx, 5.72xx, 5.72xx, 5.71xx]  # similar to CPU
```

**Actual output:**
```
[CPU]  [5.7251, 5.7227, 5.7208, 5.7193, 5.7179]
[CUDA] [6.5633, 6.5619, 6.5617, 6.5624, 6.5637]  # wrong values
```

With a small learning rate (e.g. 0.01), the raw scores barely change between iterations, making the metric appear **constant** — this causes early stopping to trigger at iteration 1.

## Verification

The CUDA-reported values match manually computing tweedie NLL using raw log-space scores as mu (confirming the missing `exp()` conversion):

```python
preds = model.predict(X_te)                      # natural space (after exp)
raw   = model.predict(X_te, raw_score=True)       # log space

def tweedie_nll(y, mu, p=1.5):
    mu = np.maximum(mu, 1e-10)
    return float(np.mean(-y * mu**(1-p)/(1-p) + mu**(2-p)/(2-p)))

print(f"Manual NLL (correct, exp preds): {tweedie_nll(y_te, preds):.4f}")  # matches CPU
print(f"Manual NLL (raw as mu, no exp):  {tweedie_nll(y_te, raw):.4f}")    # matches CUDA
```

## Root Cause

Traced through the source code:

1. **`src/objective/objective_function.cpp`**: tweedie objective has no CUDA implementation → falls back to CPU `RegressionTweedieLoss`
2. **`src/objective/regression_objective.hpp`**: `RegressionTweedieLoss` inherits `RegressionPoissonLoss` → `RegressionL2loss` → `ObjectiveFunction`. None of these **CPU** classes override `NeedConvertOutputCUDA()`
3. **`include/LightGBM/objective_function.h`**: Base `ObjectiveFunction::NeedConvertOutputCUDA()` returns `false`
4. **`src/metric/metric.cpp`**: tweedie metric **does** have a CUDA implementation `CUDATweedieMetric`
5. **`src/metric/cuda/cuda_regression_metric.cpp`**: `Eval()` checks `objective->NeedConvertOutputCUDA()` → `false` → skips `ConvertOutputCUDA()` (the `exp()` transform) → raw log-space scores passed directly to the metric kernel

Note that Poisson does **not** have this bug because it has a CUDA objective (`CUDARegressionPoissonLoss`) that correctly overrides `NeedConvertOutputCUDA() → true`.

The `gamma` objective likely has the same bug (also falls back to CPU, while `CUDAGammaMetric` exists in CUDA).

## Possible Fix

Option A: Have the CPU fallback classes (`RegressionTweedieLoss`, `RegressionGammaLoss`) override `NeedConvertOutputCUDA()` to return `true`, and implement `ConvertOutputCUDA()` with the appropriate `exp()` transform.

Option B: Fall back the tweedie/gamma **metric** to CPU as well (like `kullback_leibler` already does), so the CPU metric's `Eval()` correctly calls `ConvertOutput()`.

## Environment

- LightGBM version: 4.6.0
- CUDA version: 12.4
- OS: Ubuntu (WSL2)
- GPU: NVIDIA (CUDA-capable)

**Comment by jameslamb** [comment]
Thanks for the report.

@shiyu1994 could you comment on this when you have time?

**Comment by Shreyas-S-809** [comment]
> Thanks for the report.
> 
> [@shiyu1994](https://github.com/shiyu1994) could you comment on this when you have time?

Hey @jameslamb , i would like to work on this, could you assign me.

**Comment by jameslamb** [comment]
Thanks for your interest, but no. This is waiting for a comment from a maintainer.

If you're looking for a first contribution to the library and looking to write C/C++, we'd love help with something like #6841 or  #6703 or #6252
